runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -142,8 +152,83 @@ function M.get_namespace(client_id)
     local client = vim.lsp.get_client_by_id(client_id)
     local name = string.format('vim.lsp.%s.%d', client and client.name or 'unknown', client_id)
     _client_namespaces[client_id] = vim.api.nvim_create_namespace(name)
+    _client_namespaces_aux[client_id] = vim.api.nvim_create_namespace(name .. '.aux')
+  end
+  return _client_namespaces[client_id], _client_namespaces_aux[client_id]
+end
+
+--- @private
+--- Filter out diagnostics with tags
+--- @param diags Lsp.Diagnostic[]
+--- @return Lsp.Diagnostic[], Lsp.Diagnostic[]
+local function filter_tagged(diags)
+  ---@type Lsp.Diagnostic[], Lsp.Diagnostic[]
+  local res, filtered = {}, {}
+  for _, diag in ipairs(diags) do
+    if diag.tags and #diag.tags > 1 then
+      filtered[#filtered + 1] = diag
+    else
+      res[#res + 1] = diag
+    end
+  end
+  return res, filtered
+end
+
+local tag_to_hl = {
+  [protocol.DiagnosticTag.Unnecessary] = '@lsp.tag.unnecessary',
+  [protocol.DiagnosticTag.Deprecated] = '@lsp.tag.deprecated',
+}
+
+--- @private
+--- @param tag Lsp.DiagnosticTag
+--- @param client_id integer
+--- @return string
+local function diagnostic_tag_to_hl(tag, client_id)
+  local hl = tag_to_hl[tag]
+  if not hl then
+    vim.notify_once(
+      string.format('Unknown DiagnosticTag %d from LSP client %d', tag, client_id),
+      vim.log.levels.WARN
+    )
+    hl = 'Error'
+  end
+  return hl
+end
+
+--- @private
+--- @param bufnr integer
+--- @param client_id integer
+--- @param ns integer
+--- @param diags Lsp.Diagnostic[]
+local function highlight_tagged_diagnostics(bufnr, client_id, ns, diags)
+  vim.api.nvim_buf_clear_namespace(bufnr, ns, 0, -1)
+
+  if #diags == 0 then
+    return
+  end
+
+  vim.api.nvim_set_hl(0, '@lsp.tag.unnecessary', { link = 'Comment', default = true })
+  vim.api.nvim_set_hl(0, '@lsp.tag.deprecated', { link = '@lsp.mod.deprecated', default = true })
+
```

Comment:
Doesn't this prevent users from setting their custom highlight if this gets executed each time ?

Reply:
`default = true` should prevent that.